apps/prairielearn/src/middlewares/authzCourseOrInstance.js

```diff
@@ -96,27 +95,6 @@ module.exports = asyncHandler(async (req, res, next) => {
     res.locals.authz_data.has_student_access_with_enrollment =
       permissions_course_instance.has_student_access_with_enrollment;
     res.locals.authz_data.has_student_access = permissions_course_instance.has_student_access;
-
-    // If the user does not currently have access to the course, but could if
-    // they were enrolled, automatically enroll them. However, we will not
-    // attempt to enroll them if they are an instructor (that is, if they have
-    // a specific role in the course or course instance).
-    if (
-      res.locals.authz_data.course_role === 'None' &&
-      res.locals.authz_data.course_instance_role === 'None' &&
-      res.locals.authz_data.has_student_access &&
-      !res.locals.authz_data.has_student_access_with_enrollment
-    ) {
-      // TODO: this enrollment should enforce enrollment limits.
-      await insertEnrollment({
-        user_id: res.locals.authn_user.user_id,
-        course_instance_id: res.locals.course_instance.id,
-      });
-
-      // This is the only part of the `authz_data` that would change as a
-      // result of this enrollment, so we can just update it directly.
-      res.locals.authz_data.has_student_access_with_enrollment = true;
-    }
```

Comment:
A change in this PR that I'd very much like close attention on: I shifted auto-enrollment out of `authzCourseOrInstance` and into a separate middleware. This was necessary because I wanted to use `authzCourseOrInstance` as part of the enrollment code itself, so I couldn't have it trying to insert enrollments, if that makes sense. I'll throw a comment on the replacement middleware.